cli/src/context.rs

```diff
@@ -0,0 +1,358 @@
+    repo::BundleRepo,
+};
+use tempfile::TempDir;
+#[cfg(target_os = "macos")]
+use xcresult::XCResult;
+
+use crate::{
+    api_client::ApiClient,
+    print::print_bep_results,
+    quarantine::{run_quarantine, FailedTestsExtractor, QuarantineRunResult},
+    test::TestRunResult,
+    upload::UploadArgs,
+};
+
+pub struct PreTestContext {
+    pub meta: BundleMeta,
+    pub junit_path_wrappers: Vec<JunitReportFileWithStatus>,
+    pub bep_result: Option<BepParseResult>,
+    pub junit_path_wrappers_temp_dir: Option<TempDir>,
+}
+
+pub fn gather_pre_test_context(upload_args: UploadArgs) -> anyhow::Result<PreTestContext> {
+    let UploadArgs {
+        junit_paths,
+        #[cfg(target_os = "macos")]
+        xcresult_path,
+        bazel_bep_path,
+        org_url_slug,
+        token,
+        repo_root,
+        repo_url,
+        repo_head_sha,
+        repo_head_branch,
+        repo_head_commit_epoch,
+        tags,
+        #[cfg(target_os = "macos")]
+        allow_empty_test_results,
+        ..
+    } = upload_args;
+
+    let repo = BundleRepo::new(
+        repo_root,
+        repo_url,
+        repo_head_sha,
+        repo_head_branch,
+        repo_head_commit_epoch,
+    )?;
+
+    let (junit_path_wrappers, bep_result, junit_path_wrappers_temp_dir) =
+        coalesce_junit_path_wrappers(
+            junit_paths,
+            bazel_bep_path,
+            #[cfg(target_os = "macos")]
+            xcresult_path,
+            #[cfg(target_os = "macos")]
+            &repo.repo,
+            #[cfg(target_os = "macos")]
+            &org_url_slug,
+            #[cfg(target_os = "macos")]
+            allow_empty_test_results,
+        )?;
+
+    let envs: HashMap<String, String> = ENVS_TO_GET
+        .iter()
+        .filter_map(|&env_var| {
+            env::var(env_var)
+                .map(|env_var_value| (env_var.to_string(), env_var_value))
+                .ok()
+        })
+        .collect();
+
+    let meta = BundleMeta {
+        junit_props: BundleMetaJunitProps::default(),
+        debug_props: BundleMetaDebugProps {
+            command_line: env::args()
+                .collect::<Vec<String>>()
+                .join(" ")
+                .replace(&token, "***"),
+        },
+        bundle_upload_id_v2: String::with_capacity(0),
+        base_props: BundleMetaBaseProps {
+            version: META_VERSION.to_string(),
+            org: org_url_slug,
+            repo,
+            cli_version: format!(
+                "cargo={} git={} rustc={}",
+                env!("CARGO_PKG_VERSION"),
+                env!("VERGEN_GIT_SHA"),
+                env!("VERGEN_RUSTC_SEMVER")
+            ),
+            bundle_upload_id: String::with_capacity(0),
+            tags: parse_custom_tags(&tags)?,
+            file_sets: Vec::with_capacity(0),
+            envs,
+            upload_time_epoch: SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs(),
+            test_command: None,
```

Comment:
Are you missing a call to set `test_command`? I don't see it elsewhere...

Reply:
I totally did forget this, so I've updated it here:
[`4ea761b` (#288)](https://github.com/trunk-io/analytics-cli/pull/288/commits/4ea761bd29bbcf1c93e6ca8b37e782aba98fd12d#diff-f744e9c93b77a61a62a445a9ac43b43c353d329a0656b033aca02edc89676e5bR169)

And I added a test for that, so we don't mess up again here:
[`4ea761b` (#288)](https://github.com/trunk-io/analytics-cli/pull/288/files#diff-75cb9d34ff9cc8c9e6bdbcea3c7014f3c8f2f96d8885d780b7a989acc6446488R138-R141)